src/main/java/dev/j3fftw/litexpansion/Events.java

```diff
@@ -161,101 +156,102 @@ public void onDye(PlayerInteractEntityEvent e) {
 
         final Material blockType = block.getType();
         final Location blockLocation = block.getLocation();
+        final ItemStack hand = event.getItem();
 
         final MiningDrill miningDrill = (MiningDrill) SlimefunItem.getById(Items.MINING_DRILL.getItemId());
+        if (miningDrill.isItem(hand)) {
+            Validate.notNull(miningDrill);
+            if (!Check(miningDrill, event, blockLocation)) {
+                return;
+            }
 
-        Validate.notNull(miningDrill, "Can not be null");
-        if (miningDrill.isItem(e.getItem())
-            && SlimefunTag.STONE_VARIANTS.isTagged(blockType)
-            && !miningDrill.isDisabled()
-            && Slimefun.getProtectionManager().hasPermission(e.getPlayer(),
-            blockLocation, Interaction.BREAK_BLOCK)
-        ) {
-            e.setCancelled(true);
+            if (!SlimefunTag.STONE_VARIANTS.isTagged(blockType)) {
+                return;
+            }
 
-            final SlimefunItem slimefunItem = BlockStorage.check(block);
+            DrillUse(0.5f, block, blockType, blockLocation, event);
+        }
 
-            if (slimefunItem == null && ((Rechargeable) SlimefunItem.getByItem(e.getItem()))
-                .removeItemCharge(e.getItem(), 0.5F)
-            ) {
-                // This allows other plugins to register broken block as player broken
-                BlockBreakEvent newEvent = new BlockBreakEvent(block, e.getPlayer());
-                Bukkit.getServer().getPluginManager().callEvent(newEvent);
-                block.setType(Material.AIR);
-                e.getPlayer().playSound(blockLocation, Sound.BLOCK_STONE_BREAK, 1.5F, 1F);
-
-                // This has to be done because the item in the main hand is not a pickaxe
-                if (blockType == Material.STONE) {
-                    blockLocation.getWorld().dropItem(blockLocation,
-                        new ItemStack(Material.COBBLESTONE)
-                    );
-
-                } else {
-                    blockLocation.getWorld().dropItem(blockLocation,
-                        new ItemStack(blockType)
-                    );
-                }
+        final MiningDrill diamondDrill = (MiningDrill) SlimefunItem.getById(Items.DIAMOND_DRILL.getItemId());
+        if (diamondDrill.isItem(hand)) {
+
+            Validate.notNull(diamondDrill);
+            if (!Check(diamondDrill, event, blockLocation)) {
+                return;
             }
+
+            if (!SlimefunTag.MINEABLE_PICKAXE.isTagged(blockType)) {
+                return;
+            }
+
+            DrillUse(1.5f, block, blockType, blockLocation, event);
         }
     }
 
-    //TODO combine onDiamondDrillUse with onMiningDrillUse
-    @EventHandler
-    public void onDiamondDrillUse(PlayerInteractEvent e) {
-        final Block block = e.getClickedBlock();
-        if (block == null) {
+
+    public boolean Check(MiningDrill miningDrill, PlayerInteractEvent event, Location blockLocation) {
```

Comment:
Method naming conventions :eyes:

Reply:
whoops, i didnt see this when my friend wrote this code.

Again Thanks for the feedback <3